File: src/runner.ts
import { readFileSync, writeFileSync, existsSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { type Config } from './config.js';
import { GitLabClient } from './gitlab-client.js';
import { reviewDiff, formatReviewSummary, type ReviewResult } from './gemini-review.js';
import { fixIssues, type FixResult } from './claude-fix.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const STATE_FILE = resolve(__dirname, '..', 'state.json');

interface MrState {
  round: number;
  status: string;
}

export class StateManager {
  private state: Record<string, MrState> = {};
  private statePath: string;

  constructor(statePath = STATE_FILE) {
    this.statePath = statePath;
    this.state = this.load();
  }

  private load(): Record<string, MrState> {
    if (existsSync(this.statePath)) {
      try {
        return JSON.parse(readFileSync(this.statePath, 'utf-8'));
      } catch {
        console.warn('状态文件已损坏，重置。');
        return {};
      }
    }
    return {};
  }

  save() {
    writeFileSync(this.statePath, JSON.stringify(this.state, null, 2));
  }

  getRound(mrIid: number): number {
    return this.state[String(mrIid)]?.round ?? 0;
  }

  incrementRound(mrIid: number) {
    const key = String(mrIid);
    if (!this.state[key]) this.state[key] = { round: 0, status: 'idle' };
    this.state[key].round++;
    this.state[key].status = 'reviewing';
    this.save();
  }

  setStatus(mrIid: number, status: string) {
    const key = String(mrIid);
    if (!this.state[key]) this.state[key] = { round: 0, status };
    this.state[key].status = status;
    this.save();
  }

  isProcessing(mrIid: number): boolean {
    const status = this.state[String(mrIid)]?.status;
    return status === 'reviewing' || status === 'fixing';
  }
}

async function reviewMr(
  client: GitLabClient,
  config: Config,
  mrIid: number,
): Promise<ReviewResult> {
  console.log(`审查 MR !${mrIid}...`);

  const diffText = await client.getMrDiffText(mrIid);
  if (!diffText.trim()) {
    console.log(`MR !${mrIid} 没有变更，跳过审查。`);
    return { summary: '未检测到变更。', overall_verdict: 'approved', comments: [] };
  }

  const review = await reviewDiff(diffText, config.geminiApiKey, config.geminiModel);

  const summary = formatReviewSummary(review);
  await client.postNote(mrIid, summary);
  console.log(
    `已发布审查结果 MR !${mrIid}: 结论=${review.overall_verdict}, 问题数=${review.comments.length}`,
  );

  return review;
}

export async function runReviewFixCycle(
  client: GitLabClient,
  config: Config,
  mrIid: number,
  state: StateManager,
) {
  if (state.isProcessing(mrIid)) {
    console.log(`MR !${mrIid} 正在处理中，跳过。`);
    return;
  }

  const currentRound = state.getRound(mrIid);
  if (currentRound >= config.maxReviewRounds) {
    console.log(`MR !${mrIid} 已达最大轮数 (${config.maxReviewRounds})，停止。`);
    await client.postNote(mrIid, `⏹️ 自动审查已停止: 已达最大 ${config.maxReviewRounds} 轮。`);
    state.setStatus(mrIid, 'max_rounds_reached');
    return;
  }

  // Step 1: Review
  state.incrementRound(mrIid);
  let review: ReviewResult;
  try {
    review = await reviewMr(client, config, mrIid);
  } catch (e: unknown) {
    console.error(`MR !${mrIid} 审查失败: ${e}`);
    state.setStatus(mrIid, 'failed');
    return;
  }

  if (
    review.overall_verdict === 'approved' ||
    !review.comments.some((c) => c.severity === 'critical' || c.severity === 'warning')
  ) {
    console.log(`MR !${mrIid} 第 ${state.getRound(mrIid)} 轮审查后通过。`);
    state.setStatus(mrIid, 'approved');
    return;
  }

  // Step 2: Fix
  state.setStatus(mrIid, 'fixing');
  let fixResult: FixResult;
  try {
    fixResult = await fixIssues(config.claudeWorkDir, review, mrIid);
  } catch (e: unknown) {
    console.error(`MR !${mrIid} 修复失败: ${e}`);
    await client.postNote(mrIid, `❌ 自动修复失败: ${e}`);
    state.setStatus(mrIid, 'failed');
    return;
  }

  if (!fixResult.success) {
    console.error(`MR !${mrIid} 修复失败: ${fixResult.error}`);
    await client.postNote(mrIid, `❌ 自动修复失败: ${fixResult.error}`);
    state.setStatus(mrIid, 'failed');
    return;
  }

  if (fixResult.changed) {
    console.log(`MR !${mrIid} 已修复并推送，等待 GitLab 触发重新审查。`);
    state.setStatus(mrIid, 'fixed_waiting_for_re_review');
  } else {
    console.log(`MR !${mrIid} Claude Code 未做修改，标记为完成。`);
    state.setStatus(mrIid, 'done');
  }
}

export async function processAllOpenMrs(client: GitLabClient, config: Config) {
  const state = new StateManager();

  const openMrs = await client.listOpenMrs();
  if (openMrs.length === 0) {
    console.log('没有开放的 MR。');
    return;
  }

  for (const mr of openMrs) {
    console.log(`处理 MR !${mr.iid}: ${mr.title}`);
    await runReviewFixCycle(client, config, mr.iid, state);
    await new Promise((r) => setTimeout(r, 5000));
  }

  console.log('所有开放 MR 已处理完毕。');
}

File: src/runner.py
"""Main event loop / orchestrator for auto-review workflow."""

import json
import logging
import time
from pathlib import Path

from .config import Config, load_config
from .gitlab_client import GitLabClient
from .gemini_review import ReviewResult, review_diff, format_review_summary
from .claude_fix import fix_issues

logger = logging.getLogger(__name__)

STATE_FILE = Path(__file__).parent.parent / "state.json"


class StateManager:
    """Simple JSON file-based state tracker for MR review rounds."""

    def __init__(self, state_path: Path = STATE_FILE):
        self.state_path = state_path
        self.state = self._load()

    def _load(self) -> dict:
        if self.state_path.exists():
            try:
                with open(self.state_path) as f:
                    return json.load(f)
            except (json.JSONDecodeError, IOError):
                logger.warning(f"State file corrupted, resetting.")
                return {}
        return {}

    def save(self):
        with open(self.state_path, "w") as f:
            json.dump(self.state, f, indent=2)

    def get_round(self, mr_iid: int) -> int:
        return self.state.get(str(mr_iid), {}).get("round", 0)

    def increment_round(self, mr_iid: int):
        key = str(mr_iid)
        if key not in self.state:
            self.state[key] = {"round": 0, "status": "idle"}
        self.state[key]["round"] += 1
        self.state[key]["status"] = "reviewing"
        self.save()

    def set_status(self, mr_iid: int, status: str):
        key = str(mr_iid)
        if key not in self.state:
            self.state[key] = {"round": 0}
        self.state[key]["status"] = status
        self.save()

    def is_processing(self, mr_iid: int) -> bool:
        return self.state.get(str(mr_iid), {}).get("status") in ("reviewing", "fixing")


def review_mr(client: GitLabClient, config: Config, mr_iid: int) -> ReviewResult:
    """Step 1: Fetch MR diff, send to Gemini, post review comment."""
    logger.info(f"Reviewing MR !{mr_iid}...")

    diff_text = client.get_mr_diff_text(mr_iid)
    if not diff_text.strip():
        logger.info(f"MR !{mr_iid} has no diffs, skipping review.")
        return ReviewResult(summary="No changes detected.", overall_verdict="approved")

    review = review_diff(diff_text, config.gemini_api_key, config.gemini_model)

    # Post review summary as a note on the MR
    summary = format_review_summary(review)
    client.post_note(mr_iid, summary)
    logger.info(f"Posted review for MR !{mr_iid}: verdict={review.overall_verdict}, issues={len(review.comments)}")

    return review


def run_review_fix_cycle(client: GitLabClient, config: Config, mr_iid: int, state: StateManager):
    """Run one full review-fix cycle for an MR."""
    if state.is_processing(mr_iid):
        logger.info(f"MR !{mr_iid} is already being processed, skipping.")
        return

    current_round = state.get_round(mr_iid)
    if current_round >= config.max_review_rounds:
        logger.info(f"MR !{mr_iid} has reached max rounds ({config.max_review_rounds}), stopping.")
        client.post_note(mr_iid, f"⏹️ Auto-review stopped: reached maximum {config.max_review_rounds} rounds.")
        state.set_status(mr_iid, "max_rounds_reached")
        return

    # Step 1: Review
    state.increment_round(mr_iid)
    try:
        review = review_mr(client, config, mr_iid)
    except Exception as e:
        logger.error(f"Review failed for MR !{mr_iid}: {e}")
        state.set_status(mr_iid, "failed")
        return

    if review.overall_verdict == "approved" or not review.fixable_comments:
        logger.info(f"MR !{mr_iid} approved after round {state.get_round(mr_iid)}.")
        state.set_status(mr_iid, "approved")
        return

    # Step 2: Fix
    state.set_status(mr_iid, "fixing")
    try:
        result = fix_issues(config.claude_work_dir, review, mr_iid)
    except Exception as e:
        logger.error(f"Fix failed for MR !{mr_iid}: {e}")
        client.post_note(mr_iid, f"❌ Auto-fix failed: {e}")
        state.set_status(mr_iid, "failed")
        return

    if not result.success:
        logger.error(f"Fix failed for MR !{mr_iid}: {result.error}")
        client.post_note(mr_iid, f"❌ Auto-fix failed: {result.error}")
        state.set_status(mr_iid, "failed")
        return

    if result.changed:
        logger.info(f"MR !{mr_iid} fixed and pushed. GitLab will trigger re-review.")
        state.set_status(mr_iid, "fixed_waiting_for_re_review")
    else:
        logger.info(f"MR !{mr_iid} no changes made by Claude, marking as done.")
        state.set_status(mr_iid, "done")


def main():
    """Entry point: load config, test connection, run review-fix cycle for the latest open MR."""
    logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")

    config = load_config()
    logger.info(f"Config loaded: GitLab={config.gitlab_url}, Project={config.gitlab_project_id}")

    client = GitLabClient(config.gitlab_url, config.gitlab_token, config.gitlab_project_id)
    if not client.test_connection():
        logger.error("Cannot connect to GitLab. Check your configuration.")
        return

    state = StateManager()

    # Get all open MRs
    try:
        resp = client._request("GET", f"{client._project_path()}/merge_requests", params={"state": "opened"})
        open_mrs = resp.json()
    except Exception as e:
        logger.error(f"Failed to list MRs: {e}")
        return

    if not open_mrs:
        logger.info("No open MRs found.")
        return

    for mr in open_mrs:
        mr_iid = mr["iid"]
        logger.info(f"Processing MR !{mr_iid}: {mr['title']}")
        run_review_fix_cycle(client, config, mr_iid, state)
        # Small delay between MRs
        time.sleep(5)

    logger.info("All open MRs processed.")


if __name__ == "__main__":
    main()
